`core::marker::NoCell` in bounds (previously known an `Freeze`) (#3633)
rust-lang/rfcs: text/0000-stabilize-marker-freeze.md

The purpose of this change was to ensure that interior mutability cannot affect content that may have been static-promoted in read-only memory, which would be a soundness issue.

However, this new requirement also prevents using static-promotion to allow generics to provide a generic equivalent to `static` (with the distinction that static-promotion doesn't guarantee a unique address for the promoted content). An example of this pattern can be found in `stabby` and `equator`'s shared way of constructing v-tables:
```rust
pub trait VTable<'a>: Copy {
	const VT: &'a Self;
}
pub struct VtAccumulator<Tail, Head> {
	tail: Tail,
	head: Head,
}
impl<Tail: VTable<'a>, Head: VTable<'a>> VTable<'a> for VtAccumulator<Tail, Head> {
	const VT: &'a Self = &Self {tail: *Tail::VT, head: *Head::VT}; // Doesn't compile since 1.78
} 
```

Making `VTable` a subtrait of `core::marker::Freeze` in this example is sufficient to allow this example to compile again, as static-promotion becomes legal again. This is however impossible as of today due to `core::marker::Freeze` being restricted to `nightly`.

> Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[T; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which currently use their own equivalents of `core::marker::Freeze` to ensure that requirement, which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They would notably benefit from `core::marker::Freeze`'s status as an auto-trait.

Reviewer: ```suggestion
Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[u8; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which currently use their own equivalents of `core::marker::Freeze` to ensure that requirement, which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They would notably benefit from `core::marker::Freeze`'s status as an auto-trait.
```

(I assume that's what you meant?)

Also, re: zerocopy and bytemuck (cc @zachs18), I believe it'd be more accurate to say that:
- zerocopy is [currently developing its equivalent](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html)
- bytemuck is [considering how to use `Freeze`](https://github.com/Lokathor/bytemuck/discussions/153)

I'm not sure those distinctions are particularly important, but I figured I'd point it out in case you care about being precise. @zachs18 can comment on whether I've mis-characterized bytemuck's situation.

Finally, zerocopy doesn't need `Freeze` to be an auto-trait. It just needs it to be supported *somehow*. E.g., requiring `impl Freeze for MyType` a la `Copy` would be fine for us.

		impl S {
			pub const fn new() -> Self { S(42) }
		}
	}

	// new version of the crate, adding interior mutability.
	mod v2 {
		use std::cell::Cell;
		pub struct S(Cell<i32>);
		impl S {
			pub const fn new() -> Self { S(Cell::new(42)) }
		}
	}

	// Old version: builds
	const C1: &v1::S = &v1::S::new();
	// New version: does not build
	const C2: &v2::S = &v2::S::new();
	```

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

- The benefits of stabilizing `core::mem::Freeze` have been highlighted in [Motivation](#motivation).
	- By not stabilizing `core::mem::Freeze` in trait bounds, we are preventing useful and sound code patterns from existing which were previously supported.
- Alternatively, a non-auto sub-trait of `core::mem::Freeze` may be defined:
	- While this reduces the SemVer hazard by making its breakage more obvious, this does lose part of the usefulness that `core::mem::Freeze` would provide to projects such as `zerocopy`.
	- A "perfect" derive macro should then be introduced to ease the implementation of this trait. A lint may be introduced in `clippy` to inform users of the existence and applicability of this new trait.

> # Prior art

Reviewer: Maybe mention zerocopy's [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html) here?

Author: Double check: do you mean in the prior art section, or in the derives?

Reviewer: Prior art

Author: Done :)

The purpose of this change was to ensure that interior mutability cannot affect content that may have been static-promoted in read-only memory, which would be a soundness issue.

However, this new requirement also prevents using static-promotion to allow generics to provide a generic equivalent to `static` (with the distinction that static-promotion doesn't guarantee a unique address for the promoted content). An example of this pattern can be found in `stabby` and `equator`'s shared way of constructing v-tables:
```rust
pub trait VTable<'a>: Copy {
    const VT: &'a Self;
}
pub struct VtAccumulator<Tail, Head> {
    tail: Tail,
    head: Head,
}
impl<Tail: VTable<'a>, Head: VTable<'a>> VTable<'a> for VtAccumulator<Tail, Head> {
	const VT: &'a Self = &Self {tail: *Tail::VT, head: *Head::VT}; // Doesn't compile since 1.78
} 
```

Making `VTable` a subtrait of `core::marker::Freeze` in this example is sufficient to allow this example to compile again, as static-promotion becomes legal again. This is however impossible as of today due to `core::marker::Freeze` being restricted to `nightly`.

> Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[u8; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which are currently evaluating the use of `core::marker::Freeze` to ensure that requirement; or rolling out their own equivalents (such as zerocopy's `Immutable`) which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They could stand to benefit from `core::marker::Freeze`'s status as an auto-trait, but this isn't a requirement for them, and explicit derivation is considered suitable (at least to zerocopy).

Reviewer 2: The first time I read this, I thought it was saying that the Freeze trait is sufficient to ensure such safety. But my understanding is that it is not, since you also have to worry about padding bytes.

Author: From the [reference on undefined behaviour](https://doc.rust-lang.org/reference/behavior-considered-undefined.html):
> The following values are invalid (at their respective type):  
> An integer (i*/u*), floating point value (f*), or raw pointer obtained from uninitialized memory, or uninitialized memory in a str.  
> Note: Uninitialized memory is also implicitly invalid for any type that has a restricted set of valid values. In other words, the only cases in which reading uninitialized memory is permitted are inside unions and in "padding" (the gaps between the fields/elements of a type).

My spec-lawyer-speech is a bit wonky, but I parse this as "reading u8s from the uninitialized bytes (padding of a struct) is UB".

While I'm kind of curious if my interpretation is correct (that note leaves a hint of doubt in me); I think this is orthogonal to the purpose of `Freeze`. I'll resolve this to reduce noise on the issue :)

Reviewer 3: I think the point of the comment is that the RFC could use clarification: `Freeze` is a *part of* ensuring that the transmute is sound, but it is not sufficient.

Author: This seems slightly out of scope, but I'll add it to the RFC regardless to help future readers not get caught on this. Resolving again (the commit addressing all resolved conversations is coming soon) :)

> Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[u8; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which are currently evaluating the use of `core::marker::Freeze` to ensure that requirement; or rolling out their own equivalents (such as zerocopy's `Immutable`) which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They could stand to benefit from `core::marker::Freeze`'s status as an auto-trait, but this isn't a requirement for them, and explicit derivation is considered suitable (at least to zerocopy).

Reviewer 4: `feature(transmutability)` currently uses `Freeze` for exactly this purpose. However, we apply the `Freeze` bound recursively through references. Can someone shed some light on why `Freeze` *isn't* recursive by default?

Reviewer 3: If you have an `x: &&Cell<i32>`, then `x` itself is a perfectly `noalias` pointer without any interior mutability. `*x` can never change as long as the reference is live. Only `**x` (or rather, `(*x).get())`) can change. Therefore, `x` should be `Freeze`.

Reviewer: Also, to treat `Freeze` as recursive implies something closer to "cannot cause interior mutation anywhere", but this is misleading since even a recursively `Freeze` type could access and mutate global `static` variables.

Reviewer 3: On top of global `static, a type that is semantically equivalent to `Cell` can be implemented without `UnsafeCell` by using raw pointers.

> Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[u8; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which are currently evaluating the use of `core::marker::Freeze` to ensure that requirement; or rolling out their own equivalents (such as zerocopy's `Immutable`) which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They could stand to benefit from `core::marker::Freeze`'s status as an auto-trait, but this isn't a requirement for them, and explicit derivation is considered suitable (at least to zerocopy).

Reviewer 4: Speaking as a zerocopy maintainer: We don't consider our `Immutable` shim trait to be totally satisfactory, and we intend to replace our bespoke trait with a re-export of `core::marker::Freeze` once it is possible to do so.

```suggestion
Orthogonally to static-promotion, `core::marker::Freeze` can also be used to ensure that transmuting `&T` to a reference to an interior-immutable type (such as `[u8; core::mem::size_of::<T>()]`) is sound (as interior-mutation of a `&T` may lead to UB in code using the transmuted reference, as it expects that reference's pointee to never change). This is notably a safety requirement for `zerocopy` and `bytemuck` which are currently evaluating the use of `core::marker::Freeze` to ensure that requirement; or rolling out their own equivalents (such as zerocopy's `Immutable`) which imposes great maintenance pressure on these crates to ensure they support as many types as possible. They could stand to benefit from `core::marker::Freeze`'s status as an auto-trait, and `zerocopy` intends to replace its bespoke trait with a re-export of `core::marker::Freeze` .
```

It is automatically implemented by the compiler for any type that doesn't shallowly contain a `core::cell::UnsafeCell`.

Notably, a `const` can only store a reference to a value of type `T` if `T: core::marker::Freeze`, in a pattern named "static-promotion".

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The following documentation is lifted from the current nightly documentation.
```markdown
Used to determine whether a type contains
any `UnsafeCell` internally, but not through an indirection.
This affects, for example, whether a `static` of that type is
placed in read-only static memory or writable static memory.
This can be used to declare that a constant with a generic type
will not contain interior mutability, and subsequently allow
placing the constant behind references.
# Safety
This trait is a core part of the language, it is just expressed as a trait in libcore for
convenience. Do *not* implement it for other types.
```

The current _Safety_ section may be removed, as manual implementation of this trait is forbidden.

> From a cursary review, the following documentation improvements may be considered:

Reviewer 5: ```suggestion
From a cursory review, the following documentation improvements may be considered:
```

```markdown
[`Freeze`](core::marker::Freeze) marks all types that do not contain any un-indirected interior mutability.
This means that their byte representation cannot change as long as a reference to them exists.

Note that `T: Freeze` is a shallow property: `T` is still allowed to contain interior mutability,
provided that it is behind an indirection (such as `Box<UnsafeCell<U>>`).

> Notable interior mutability sources are [`UnsafeCell`](core::cell::UnsafeCell) (and any of its safe wrappers
> such the types in the [`cell` module](core::cell) or [`Mutex`](std::sync::Mutex)) and [atomics](core::sync::atomic). 

Reviewer 6: `core::sync::atomic::Atomic*` are _also_ safe wrappers of `UnsafeCell`, not something different. How about this?

```suggestion
Notable `!Freeze` types are [`UnsafeCell`](core::cell::UnsafeCell) and its safe wrappers
such as the types in the [`cell` module](core::cell), [`Mutex`](std::sync::Mutex)), and [atomics](core::sync::atomic).
Any type which contains any one of these without indirection is also `!Freeze`.
```

`T: Freeze` is notably a requirement for static promotion (`const REF: &'a T;`) to be legal.

Note that static promotion doesn't guarantee a single address: if `REF` is assigned to multiple variables,
they may still refer to distinct addresses.

Whether or not `T: Freeze` may also affect whether `static STATIC: T` is placed
in read-only static memory or writeable static memory, or the optimizations that may be performed
in code that holds an immutable reference to `T`.
```

> Mention could be added to `UnsafeCell` and atomics that adding one to a previously `Freeze` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoque its implementation of `Freeze`.

Reviewer 5: ```suggestion
Mention could be added to `UnsafeCell` and atomics that adding one to a previously `Freeze` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoke its implementation of `Freeze`.
```

# Drawbacks
[drawbacks]: #drawbacks

- Some people have previously argued that this would be akin to exposing compiler internals.
	- The RFC author disagrees, viewing `Freeze` in a similar light as `Send` and `Sync`: a trait that allows soundness requirements to be proven at compile time.
- `Freeze` being an auto-trait, it is, like `Send` and `Sync` a sneaky SemVer hazard.

> 	- Note that this SemVer hazard already exists through the existence of static-promotion, as examplified by the following example:

Reviewer 5: ```suggestion
	- Note that this SemVer hazard already exists through the existence of static-promotion, as exemplified by the following example:
```

	```rust

> 	// old version of the crate.
> 	mod v1 {
> 		pub struct S(i32);
> 		impl S {
> 			pub const fn new() -> Self { S(42) }
> 		}
> 	}
> 
> 	// new version of the crate, adding interior mutability.
> 	mod v2 {
> 		use std::cell::Cell;
> 		pub struct S(Cell<i32>);
> 		impl S {
> 			pub const fn new() -> Self { S(Cell::new(42)) }
> 		}
> 	}
> 
> 	// Old version: builds
> 	const C1: &v1::S = &v1::S::new();
> 	// New version: does not build
> 	const C2: &v2::S = &v2::S::new();

Reviewer 5: ```suggestion
    // old version of the crate.
    mod v1 {
        pub struct S(i32);
        impl S {
            pub const fn new() -> Self { S(42) }
        }
    }

    // new version of the crate, adding interior mutability.
    mod v2 {
        use std::cell::Cell;
        pub struct S(Cell<i32>);
        impl S {
            pub const fn new() -> Self { S(Cell::new(42)) }
        }
    }

    // Old version: builds
    const C1: &v1::S = &v1::S::new();
    // New version: does not build
    const C2: &v2::S = &v2::S::new();
```

	```

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

- The benefits of stabilizing `core::mem::Freeze` have been highlighted in [Motivation](#motivation).
	- By not stabilizing `core::mem::Freeze` in trait bounds, we are preventing useful and sound code patterns from existing which were previously supported.
- Alternatively, a non-auto sub-trait of `core::mem::Freeze` may be defined:
	- While this reduces the SemVer hazard by making its breakage more obvious, this does lose part of the usefulness that `core::mem::Freeze` would provide to projects such as `zerocopy`.
	- A "perfect" derive macro should then be introduced to ease the implementation of this trait. A lint may be introduced in `clippy` to inform users of the existence and applicability of this new trait.

# Prior art
[prior-art]: #prior-art

> - This feature has been available in `nightly` for 7 years, and is used internally by the compiler.

Reviewer 3: I don't think it is correct that this has existed in nightly for 7 years. What is the source of that claim? It exists in nightly since https://github.com/rust-lang/rust/pull/121840, which is just a few months ago.
```suggestion
- This trait has a long history: it existed in ancient times but got [removed](https://github.com/rust-lang/rust/pull/13076) before Rust 1.0.
  In 2017 it got [added back](https://github.com/rust-lang/rust/pull/41349) as a way to simplify the implementation of the `interior_unsafe` query, but it was kept private to the standard library.
  In 2019, a [request](https://github.com/rust-lang/rust/issues/60715) was filed to publicly expose the trait, but not a lot happened until recently when the issue around static promotion led to it being [exposed unstably](https://github.com/rust-lang/rust/pull/121840).
```

- The work necessary for this RFC has already been done and merged in [this PR](https://github.com/rust-lang/rust/issues/121675), and a [tracking issue](https://github.com/rust-lang/rust/issues/121675) was opened.
- zerocopy's [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html) seeks to provide the same guarantees as `core::marker::Freeze`.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> [Should the trait be exposed under a different name?](https://github.com/rust-lang/rust/pull/121501#issuecomment-1962900148)

Reviewer 6: Something that seems troublesome about `Freeze` is that it's a trait named using a verb, which is good in isolation, but that verb would usually be an _operation you can perform_ on values of the implementing type. What's the “freeze” operation on `T: Freeze`? I suppose it's taking a `&` reference to `T`. That's reasonable enough, _but_ there is another “freeze” operation that's both already named in LLVM and [recently been proposed for addition to Rust](https://github.com/rust-lang/rfcs/pull/3605): “freezing” uninitialized memory to enable safely reading the bytes of a value that contains padding, or reading a value that may be partially initialized. That's a *different thing in the same problem domain* (what you can or can't do with the bytes when you have a `T` or `&T`) which seems dangerously confusing. And outside of that name collision, “`Freeze`” doesn't convey much to the naïve reader.

Unfortunately, I don't have any _good_ ideas for alternatives, especially not [non-negated ones as @RalfJung suggested](https://github.com/rust-lang/rust/pull/121501#issuecomment-1962900148), but I think that stabilizing this under the name `Freeze` would create a frequent source of confusion among those encountering it for the first time. It might be better to have an awkward complex name than a cryptic simple one. Awkward suggestion: `ShallowImmut`?

Reviewer 7: what about `Stuck`? that is unused afaict so people seeing it would think to look up what it means in Rust. It's also nice and short and reminds users of what it means.

Reviewer 3: At least on the theory side, a program being "stuck" has meaning already -- it means that execution can't progress further as the program has left the realm of well-defined executions.

Reviewer 7: > At least on the theory side, a program being "stuck" has meaning already -- it means that execution can't progress further as the program has left the realm of well-defined executions. I would find this term very confusing.

I was thinking more of that the value in memory is stuck at whatever bytes are put into `.rodata`. because `Stuck` is a trait and applies to types, not program execution, it seems different enough to me that it won't be that confusing.

> Maybe we shouldn't call that "Freeze"? It's not a great name as what the operation really does is turn poison into non-deterministic regular bytes. Rust doesn't have undef-style "bytes that are unstable".

since the LLVM IR op is already called freeze, they've effectively already laid claim to that name, so anyone who's used LLVM IR and tries to figure out rust will be confused if we use `freeze` to mean something completely different where it could be confused with an operation (since freeze is present-tense, unlike stuck where being past-tense and being applied to types makes it much less likely to be an operation).

Reviewer 3: > I was thinking more of that the value in memory is stuck at whatever bytes are put into .rodata

That's far from it, though. `Freeze` isnt just about `rodata`, it's about all shared references without interior mutability.

I find "stuck" extremely unintuitive. It conveys negative feelings to me, "we're stuck somewhere or with something (and want to get unstuck again)". I don't think it is a good term for "this memory is immutable".

> since the LLVM IR op is already called freeze, they've effectively already laid claim to that name, so anyone who's used LLVM IR and tries to figure out rust will be confused if we use freeze to mean something completely different where it could be confused with an operation (since freeze is present-tense, unlike stuck where being past-tense and being applied to types makes it much less likely to be an operation).

LLVM also uses poison/undef, neither of which we use in Rust (we call it "uninitialized memory" or just "uninit"). LLVM calls it `getelementptr` or `ptraddr`, we call it `offset`. LLVM calls it load/store, we call it read/write. The list can probably go on for a while. I don't think we should let LLVM dictate our terminology.

Reviewer 7: > LLVM also uses poison/undef, neither of which we use in Rust (we call it "uninitialized memory" or just "uninit"). LLVM calls it `getelementptr` or `ptraddr`, we call it `offset`. LLVM calls it load/store, we call it read/write. The list can probably go on for a while. I don't think we should let LLVM dictate our terminology.

yes, I have no problem using a different name than LLVM, what I have a problem with is using the same name as LLVM but for a different operation. (yes, I think naming `f32::muladd` was a mistake).

Reviewer: In zerocopy, we went with [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.12/zerocopy/trait.Immutable.html), which I quite like (credit to @jswrenn's coworker for suggesting the name). It is an adjective and accurately describes the semantics - `T: Immutable` means that `T` cannot be mutated behind a shared reference.

Reviewer 7: but `T` can be mutated behind a shared reference if that mutation is hidden in some other block of memory: `Box<Cell<u8>>` does implement `Freeze`, but can be mutated behind a shared reference. therefore I think using `Immutable` would be highly misleading.

Reviewer: Arguably *any* name that doesn't clarify that the property only applies shallowly/not-via-indirection is similarly misleading. If we're worried about that, we'd need to add an adjective like `Shallow` to the name.

Reviewer 8: Maybe just invent another term that no one else used. `Freeze` itself is an invented term anyway. 

Perhaps `Solid` (you "freeze" liquid to get "solid" :shrug:). 
Or `Acellular` (technically a negated word).

But given that the trait has existed for so long under the name `Freeze` perhaps it has already gotten traction 

Though IMO it's fine to actually use a multi-word term like `ShallowImmutable` or `CellFree`? We already have got the the two-word `UnwindSafe` and three-word `RefUnwindSafe` auto traits in std. As long as this trait bound is not needed very often (unlike `Send` and `Sync` and `Sized`) using a longer but more precise name shouldn't be a problem.

Author: Let me set a record for the most negative name: `UnindirectedUnUnsafeCelled` :smile: 

I like `ShallowImmutable` a lot, as it's rather descriptive and attracts curiosity thanks to the `Shallow` term.

Do we have any opposition to `ShallowImmutable`?

# Future possibilities
[future-possibilities]: #future-possibilities

> - One might later consider whether `core::mem::Freeze` should be allowed to be `unsafe impl`'d like `Send` and `Sync` are, possibly allowing wrappers around interiorly mutable data to hide this interior mutability from constructs that require `Freeze` if the logic surrounding it guarantees that the interior mutability will never be used.

Reviewer 9: If extern types ever get stabilised they would present a situation where `unsafe impl Freeze` would be useful. Extern types represent types opaque to the Rust compiler (usually because they're defined in another language and accessed via FFI) and so they have to be `!Freeze` by default. I'm not sure that's enough motivation to change the approach for this RFC, but it does present a much simpler and less scary reason to want to implement Freeze for a type

Reviewer 8: An `extern type T` can never appear directly in a struct layout, you must use indirection like `*const T` and `&T`, but a reference/pointer itself is `Freeze`, so whether `T: Freeze` or not does not really matter.

Regarding the motivation (to properly address rust-lang/rust#121250, well you can't dereference a `T`-pointer to create a value `T` in Rust, so the motivation doesn't apply to extern type at all.

Reviewer 9: Ahhh that's a very good point, don't mind me just gonna go steal that and write it in the extern types RFC. 😁

Reviewer 7: > An `extern type T` can never appear directly in a struct layout, you must use indirection like `*const T` and `&T`, but a reference/pointer itself is `Freeze`, so whether `T: Freeze` or not does not really matter.

I hope Rust will support extern types in structs in the future, e.g.:

```rust
extern {
    pub type Body;
}

unsafe impl Aligned for Body {
    const ALIGN: usize = 8;
}

#[repr(C)]
pub struct MyStruct {
    tag: u8,
    body: Body, // fine since rust knows what alignment is needed
}
```

Reviewer 8: @programmerjake In what scenario such a `MyStruct` will appear? AFAIK you can't use an incomplete type in C or C++.

```c
#include <stdint.h>

typedef struct Body Body;

typedef struct MyStruct {
    uint8_t tag;
    _Alignas(8) Body body;
    //^~ error: field 'body' has incomplete type
} MyStruct;
```

Reviewer 7: > @programmerjake In what scenario such a `MyStruct` will appear? AFAIK you can't use an incomplete type in C or C++.

In C it would be written something like:
```c
#include <stdint.h>

typedef struct MyStruct {
    uint8_t tag;
    _Alignas(8) char body[]; // always cast to pointer to whatever actual body type is used before use
} MyStruct;
```
or
```c
typedef struct MyStruct {
    uint8_t tag;
    char padding[7];
    // _Alignas(8) T body;
} MyStruct;
```

Reviewer 7: an example of how it's done in C is Vulkan's [`VkBaseInStructure`](https://github.com/KhronosGroup/Vulkan-Headers/blob/5677bafb820e476441e9e1f745371b72133407d3/include/vulkan/vulkan_core.h#L2856-L2859) struct, which is a prefix of basically every struct used in Vulkan's API.

Reviewer 3: This is the Freeze RFC, not the extern type RFC. :)

I don't think this RFC has to take into account extensions to unstable features.

Reviewer 8: @programmerjake if you need a `char[]` on the C side you should also use a "thin slice" on the Rust side.

While `VkBaseInStructure` is indeed the "base class" of many other Vk classes, I don't think it is appropriate to insert the "tail" into `VkBaseInStructure` itself. (Typical proposal for Rust to support OOP are like #2431) Besides, even if you have that "extern type tail" you can't realize a `VkBaseInStructure` as a concrete subclass type like `VkMemoryBarrier` anyway.

Author: I'd argue this purpose would be much better satisfied using a trait (the sacrilegious maniac in me wants to say `Deref` would even do the trick) that lets you access that "base class" (either by casting the reference or by calling a library-defined function, such that the offset may be changed by the library if suitable), rather than introduce a language feature that wouldn't really bring much to the table.

Regardless, this has little to do with this RFC, as the status of `unsafe impl Freeze` is _deliberately_ left out of scope. I'll resolve this conversation to reduce noise :)

> - One might later consider whether `core::mem::Freeze` should be allowed to be `unsafe impl`'d like `Send` and `Sync` are, possibly allowing wrappers around interiorly mutable data to hide this interior mutability from constructs that require `Freeze` if the logic surrounding it guarantees that the interior mutability will never be used.

Reviewer 3: I don't think this is a plausible possibility. If you never use interior mutability, just don't use a type with UnsafeCell. What is the possible use-case for having an UnsafeCell that you will *definitely never use for interior mutability*? (That would be the only situation where `unsafe impl Freeze` would be sound.)

Reviewer: We've considered wanting this for zerocopy. The use case would be to add a wrapper type `Frozen<T>` (joking - no clue what we'd name it) that is `Freeze` even if `T` isn't. IIUC this would require `Freeze` to be a runtime property rather than a type property (ie, interior mutation never happens), and would require being able to write `unsafe impl<T> Freeze for Frozen<T>`.

Reviewer 6: > What is the possible use-case for having an UnsafeCell that you will definitely never use for interior mutability?

Speculation: Perhaps a data structure whose contents are computed using interior mutability, then put in a wrapper struct which disables all mutation operations and implements `Freeze`. The reason to do this using a wrapper rather than transmuting to a non-interior-mutable type would be to avoid layout mismatches due to `UnsafeCell<T>` not having niches that `T` might have.

This is only relevant if such a type _also_ wants to support the functionality enabled by implementing `Freeze`, of course.

Reviewer 3: > We've considered wanting this for zerocopy. The use case would be to add a wrapper type Frozen<T> (joking - no clue what we'd name it) that is Freeze even if T isn't.

What would that type be good for?

Reviewer: > > We've considered wanting this for zerocopy. The use case would be to add a wrapper type Frozen (joking - no clue what we'd name it) that is Freeze even if T isn't.
> 
> What would that type be good for?

It's been a while since we considered this design, so the details are hazy in my memory, but roughly we wanted to be able to validate that a `&[u8]` contained the bytes of a bit-valid `&T`. We wanted to encode in the type system that we'd already validated size and alignment, so we wanted a type that represented "this has the size and alignment of `T` and all of its bytes are initialized, but might not be a bit-valid `T`." To do this, we experimented with a `MaybeValid<T>` wrapper.

When you're only considering by-value operations, you can just do `#[repr(transparent)] struct MaybeValid<T>(MaybeUninit<T>);` (`MaybeUninit` itself isn't good enough because it doesn't promise that its bytes are initialized - the newtype lets you add internal invariants). The problem is that this doesn't work by reference - if you're using `Freeze` as a bound to prove that `&[u8] -> &MaybeValid<T>` is sound, it won't work if `T: !Freeze`.

As I said, we didn't end up going that route - instead of constructing a `&MaybeValid<T>`, we construct (roughly) a `NonNull<T>` (okay, actually a [`Ptr`](https://google.github.io/zerocopy/zerocopy/pointer/ptr/def/struct.Ptr.html)) and just do the operations manually. The code in question starts at [this API](https://docs.rs/zerocopy/0.8.0-alpha.14/zerocopy/trait.TryFromBytes.html#method.try_ref_from) if you're curious.

Reviewer 3: If you're wrapping `T` inside a private field, why does it matter whether it has interior mutability or not?

The only possible benefit I can see here would be still getting `noalias` optimizations for `Frozen<T>`, but you're not mentioning optimizations. In terms of soundness, if making things less `Freeze` can never make code unsound. (https://github.com/rust-lang/unsafe-code-guidelines/issues/303 is a Stacked Borrows bug.)

Reviewer: > If you're wrapping `T` inside a private field, why does it matter whether it has interior mutability or not?

Because we want to be able to use `MaybeValid<T>` with APIs that are safe and use trait bounds (like `Freeze`) to prove their soundness internally. One of our golden rules for developing zerocopy is to be extremely anal about *internal* abstraction boundaries [1]. While we could just do what you're suggesting using `unsafe` (namely, use `unsafe` to directly do the `&[u8] -> &MaybeValid<T>` cast), we have existing internal APIs such as [this one](https://google.github.io/zerocopy/zerocopy/pointer/ptr/def/struct.Ptr.html#method.try_cast_into) that permit this conversion safely (ie, the method itself is safe to call), using trait bounds to enforce soundness.

In general, we've found that, as we teach our APIs to handle more and more cases, it quickly becomes very unwieldy to program "directly" using one big `unsafe` block (or a sequence of `unsafe` blocks) in each API. Decomposition using abstractions like `Ptr` has been crucial to us being confident that the code we're writing is actually sound.

[1] See "Abstraction boundaries and unsafe code" in our [contributing guidelines](https://github.com/google/zerocopy/discussions/1318)

Reviewer 3: I'm afraid it is not clear to me how `Frozen` helps build up internal abstraction barriers, and your existing codebase is way too big for me to be able to extract that kind of information from it with reasonable effort.

Reviewer: *TLDR: We can only make this code compile if `MaybeValid<T>: Freeze` despite `T: ?Freeze`. I'm arguing that, in order to support this use case, we should design the RFC to be forwards-compatible with permitting unsafe `impl<T> Freeze for MaybeValid<T>` even if we don't support it out of the gate.*

[Here's a minification](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ce0b446f00f2fad293f4f1c8e64d40ba) of the sort of code I'm describing. It has three components that, in reality, would live in unrelated parts of our codebase (they wouldn't be right next to each other like they are here, so we'd want to be able to reason about their behavior orthogonally):
- `MaybeValid<T>`
- `try_cast_into<T>(bytes: &[u8]) -> Option<&MaybeValid<T>> where MaybeValid<T>: Freeze`
  - This is responsible for checking size and alignment, but not validity; in our codebase, it is used both by `FromBytes` (which needs no validity check) and by `TryFromBytes` (which performs a validity check after casting)
  - In our codebase (but not in this example), this does some complicated math to compute pointer metadata for slice DST; it can't easily be inlined into its caller
- `TryFromBytes`, with:
  -  `try_ref_from(bytes: &[u8]) -> Option<&Self> where Self: Freeze`
  -  `try_mut_from(bytes: &mut [u8]) -> Option<&mut Self>`

Notice the `MaybeValid<T>: Freeze` bound on `try_cast_into` (and the safety comments inside that function). That bound permits us to make the function safe to call (without that bound, we'd need a safety precondition about whether interior mutation ever happens using the argument/returned references).

`try_ref_from` can satisfy that bound because it already requires `Self: Freeze`, which means that `MaybeValid<Self>: Freeze`. However, `try_mut_from` does not require `Self: Freeze`. Today, the linked code fails to compile thanks to line 114 in `try_mut_from`:

```rust
let maybe_self = try_cast_into::<Self>(bytes)?;
```

What we'd like to do is be able to guarantee that interior mutation will never be exercised via `&MaybeValid<T>`, and thus be able to write `unsafe impl<T> Freeze for MaybeValid<T>`.

Reviewer 3: > Notice the MaybeValid<T>: Freeze bound on try_cast_into (and the safety comments inside that function). That bound permits us to make the function safe to call (without that bound, we'd need a safety precondition about whether interior mutation ever happens using the argument/returned references).

What goes wrong if it's not Freeze? There can't even be a proof obligation attached to this since you want it to be trivially true. I think all you'd lose is some optimization potential.

If `MaybeValid` was always `Freeze`, you'd still have to extend the safety comments on `deref_unchecked` to say "also you must never ever mutate via this shared reference, even if `T` is e.g. `Cell<T>`".

Given that you need such a safety requirement anyway, MaybeValid is already a type that effectively permits no mutation of its interior ever. It then follows that you can drop the `Freeze` bound on `try_cast_into`. This requires `try_cast_into` to rely on a property of `MaybeValid`, but since `MaybeValid` appears in `try_cast_into`'s signature, there's no new coupling here -- you *already* depend on the fact that MaybeValid can be soundly constructed for invalid data.

So I still don't see a motivation for `impl Freeze` here, except if you somehow want more optimizations on `MaybeValid`.

Reviewer: I see your point, and I think that's a compelling argument. I can imagine there being cases where you'd want to pass `MaybeValid` to a generic API that accepted `T: Freeze`, in which case you'd still need this, but I don't have a concrete example of this off the top of my head.

I'd still advocate for keeping this listed as a future possibility since doing so doesn't have any effect on the current proposal (ie, the current proposal is already forwards-compatible with this). Obviously we'd need to adjudicate it much more thoroughly if it was ever proposed to actually go through with permitting `unsafe impl Freeze`, and I assume you'd push back if that ever happened just as you are now, but I think it's at least worth mentioning that it's something that we've considered and that there might be some desire for.

convenience. Do *not* implement it for other types.
```

Mention could be added to `UnsafeCell` and atomics that adding one to a previously `Freeze` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoke its implementation of `Freeze`.

## `core::marker::PhantomNotFreeze`

This ZST is proposed as a means for maintainers to reliably opt out of `Freeze` without constraining currently `!Freeze` ZSTs to remain so. While the RFC author doesn't have the expertise to produce its code,
here's its propsed documentation:

```markdown
[`PhantomNotFreeze`] is type with the following guarantees:
- It is guaranteed not to affect the layout of a type containing it as a field.
- Any type including it in its fields (including nested fields) without indirection is guaranteed to be `!Freeze`.

This latter property is [`PhantomNotFreeze`]'s raison-d'être: while other Zero-Sized-Types may currently be `!Freeze`,
[`PhantomNotFreeze`] is the only ZST that's guaranteed to keep that bound.

Notable types that are currently `!Freeze` but might not remain so in the future are:
- `UnsafeCell<T>` where `core::mem::size_of::<T>() == 0` 
- `[T; 0]` where `T: !Freeze`.

> Note that `core::marker::PhantomData<T>` if `Freeze` regardless of `T`'s `Freeze`ness.

Reviewer 10: ```suggestion
Note that `core::marker::PhantomData<T>` is `Freeze` regardless of `T`'s `Freeze`ness.
```

- zerocopy's [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html) seeks to provide the same guarantees as `core::marker::Freeze`.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- [Should the trait be exposed under a different name?](https://github.com/rust-lang/rust/pull/121501#issuecomment-1962900148)
	- An appealing proposition is `ShallowImmutable` to avoid collision with `llvm`'s `freeze`, while highlighting that the property is "shallow".

# Future possibilities
[future-possibilities]: #future-possibilities

- One might later consider whether `core::mem::Freeze` should be allowed to be `unsafe impl`'d like `Send` and `Sync` are, possibly allowing wrappers around interiorly mutable data to hide this interior mutability from constructs that require `Freeze` if the logic surrounding it guarantees that the interior mutability will never be used.
	- The current status-quo is that it cannot be implemented manually (experimentally verified with 2024-05-12's nightly).

> 	- The RFC author is unable to tell whether allowing manual implementation may cause the compiler to generate unsound code (even if the wrapper correctly prevents interior mutation), but judges that the gains of allowing these implementations are too small to justify allowing the risk.

Reviewer 3: ```suggestion
	- An `unsafe impl Freeze for T` would have very subtle soundness constraints: with such a declaration, performing mutation through any `&T` *or any pointer derived from it* would be UB. So this completely disables any interior mutability on fields of `T` with absolutely no way of ever recovering mutability.
	- Given these tight constraints, it is unclear what a concrete use-case for `unsafe impl Freeze` would be. So far, none has been found.
```

# Summary
[summary]: #summary

- Stabilize `core::marker::Freeze` in trait bounds.
	- Rename `core::marker::Freeze` to `core::marker::ShallowImmutable`. This proposition is tentative, the RFC will keep on using the historical `core::marker::Freeze` name.
	- Provide a marker type to opt out of `core::marker::Freeze` for the most semver-conscious maintainers. Tentatively named `core::marker::PhantomNotFreeze` (or `core::marker::PhantomNotShallowImmutable` to go with the proposed rename)

# Motivation
[motivation]: #motivation

With 1.78, Rust [changed behavior](https://github.com/rust-lang/rust/issues/121250): previously, `const REF: &T = &expr;` was (accidentally) accepted even when `expr` may contain interior mutability.
Now this requires that the type of `expr` satisfies `T: core::marker::Freeze`, which indicates that `T` doesn't contain any un-indirected `UnsafeCell`, meaning that `T`'s memory cannot be modified through a shared reference.

The purpose of this change was to ensure that interior mutability cannot affect content that may have been static-promoted in read-only memory, which would be a soundness issue.

> However, this new requirement also prevents using static-promotion to create constant references to data of generic type. This pattern can be used to approximate "generic `static`s" (with the distinction that static-promotion doesn't guarantee a unique address for the promoted content). An example of this pattern can be found in `stabby` and `equator`'s shared way of constructing v-tables:
> However, this new requirement also prevents using static-promotion to allow generics to provide a generic equivalent to `static` (with the distinction that static-promotion doesn't guarantee a unique address for the promoted content). An example of this pattern can be found in `stabby` and `equator`'s shared way of constructing v-tables:

Reviewer 11: ```suggestion
However, this new requirement also prevents using static-promotion to create constant references to data of generic type. This pattern can be used to approximate "generic `static`s" (with the distinction that static-promotion doesn't guarantee a unique address for the promoted content). An example of this pattern can be found in `stabby` and `equator`'s shared way of constructing v-tables:
```

(IINM, this nearly-duplicated text was left over after applying another suggestion)

[`PhantomData<T>`](core::marker::PhantomData) only implements `Freeze` if `T` does, making it a good way
to conditionally remove a generic type's `Freeze` auto-impl.

# Safety
This trait is a core part of the language, it is just expressed as a trait in libcore for
convenience. Do *not* implement it for other types.
```

Mention could be added to `UnsafeCell` and atomics that adding one to a previously `Freeze` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoke its implementation of `Freeze`.

## Fixing `core::marker::PhantomData`'s `Freeze` impl

At time of writing, `core::marker::PhantomData<T>` implements `Freeze` regardless of whether or not `T` does.

This is now considered a bug, with the corrected behaviour being that `core::marker::PhantomData<T>` only implements `Freeze` if `T` does.

While crates that would "observe" this change exist, the current consensus is that it would only break invalid usages of that invariable bound.
Most crates that would observe this change could replace their usage of `core::marker::PhantomData<T>` by `core::marker::PhantomData<Ptr<T>>` where `Ptr<T>`

> is a pointer-type with the relevant caracteristics regarding lifetime, `Send` and `Sync`ness.

Reviewer 12: The cases affected by this change seem like quite the edge-cases, and the fact that there _are_ workarounds makes me personally not super worried about it from a _breakage_ standpoint[^3]. Also, AFAICT this change shouldn't lead to any *miscompilations* since it can only *remove* `Freeze` from types that are currently `Freeze`; so the "worst kind" of breakage is completely avoided.

The necessary workarounds for code affected by this change should probably be documented somewhere, though. Perhaps just the release notes of when this change happens.

-----

As for the "necessary workarounds":

`Box<T>` works for "ptr with the same auto-traits as `T` except is `Freeze`"[^1] (though see below for `Unpin`), but note that in `#![no_std]` code without `extern crate alloc;`, there is no "pre-built" pointer type in `core` for this:
* `&mut T`: invariant in `T`, might necessitate adding a lifetime parameter where there wasn't one previously(?), 
* `&T`: wrong `Send`ness (`&T: Send` when `T: Sync`), might necessitate adding a lifetime parameter where there wasn't one previously(?)
* `*mut T`: invariant in `T`, wrong `Send`/`Sync`ness (never `Send` or `Sync`)
* `*const T`/`NonNull<T>`: wrong `Send`/`Sync`ness (never `Send` or `Sync`)

However, users can use `PhantomData<&T/*const T/NonNull<T>>` and just manually `impl Send/Sync`[^2] as appropriate to fix the differences if they are *really* constrained and cannot even use `extern crate alloc;`.

-----

For `Unpin`, all of these pointer types (including `Box<T>`) are unconditionally `Unpin`, so if users *really* need to keep the "`MyType<T>: Unpin` iff `T: Unpin`" behavior while *also* remaining `Freeze`, they'd need to also add a `PhantomPinned` field and an `impl<T: Unpin> Unpin for MyType<T>`[^4].

[^1]: I am completely ignoring `UnwindSafe` and `RefUnwindSafe` and have not checked them for anything in this comment.
[^2]: If they use `PhantomData<&T>`, then they only need to `impl<T: Send> Send for MyType<T>`[^4], as the `Sync` impl is already correct. If they use `PhantomData<*const T/NonNull<T>>`, they'd need to have manual impls for both `Send` and `Sync` to get back the original behavior in the general case, but wouldn't need to add a lifetime parameter.

[^3]: I'll put any "is this the best choice of behavior" comments in a different thread

[^4]: modified as necessary to account for other fields

This latter property is [`PhantomNotFreeze`]'s raison-d'être: while other Zero-Sized-Types may currently be `!Freeze`,
[`PhantomNotFreeze`] is the only ZST (outside of [`PhantomData<T>`] where `T` isn't `Freeze`) that is guaranteed to stay that way.

Notable types that are currently `!Freeze` but might not remain so in the future are:
- `UnsafeCell<T>` where `core::mem::size_of::<T>() == 0`
- `[T; 0]` where `T: !Freeze`.
```

This new marker type shall be introduced at the same time as the stabilization of `Freeze` in bounds.

## Addressing the naming

A point of contention during the RFC's discussions was whether `Freeze` should be renamed, as `freeze` is already a term used in `llvm` to refer to an intrinsic which allows to safely read from uninitialized memory.
[Another RFC](https://github.com/rust-lang/rfcs/pull/3605) is currently open to expose this intrinsic in Rust.

Debates have landed on the conservation of the `Freeze` name, under the main considerations that:
- No better name was found despite the efforts in trying to find one,
- that the current name was already part of the Rust jargon,

> - and that stabilizing this feature too valuable to hold it back on naming.

Reviewer: I'd like to push back on this. Our experience in zerocopy has shown that `Immutable` (the name for [our trait](https://docs.rs/zerocopy/0.8.23/zerocopy/trait.Immutable.html) whose semantics are ~identical to `Freeze`) is a very appropriate name. We often end up writing safety comments/proofs which need to reference the concept of interior mutability, and `Immutable` has proven a very intuitive name for this. I'd suggest that `Immutable` would be a much better name than `Freeze` for stabilization.

Author: This proposition was made several times during the development of this RFC.

I'm strongly against "just" `Immutable`, as it straight up lies and could give false impressions at a glance.

Reviewer: What is the sense in which it lies? That `&mut T` still permits mutation?

Reviewer 4: I'd also like to voice serious reservations about this decision:
1. `Freeze` is definitely *not* part of the **average** Rust programmer's lexicon — it's compiler jargon.
2. `Freeze`, as noted in the preceding paragraph, conflicts with an entirely unrelated concept. 

Naming is typically finalized during stabilization, and I'm surprised that this feature — unlike all other library additions — is considered so urgent as to warrant a fast-tracked stabilization that bypasses the usual naming discussions. Naming this trait is the one shot we get at picking what term *should* be in the average Rust programmer's lexicon, and I don't think we should discard that opportunity lightly.

Author: Exactly. One could expect that at first glance `Immutable` is somehow protected from mutation even if you have the sole reference to it (as people like to state that `&uniq` is a more "exact" name for what `&mut` truly is).

Too simple a name would give an illusion of simplicity which may turn into a trap. Some languages are built around the concept of immutability, so it's not far fetch that someone may find `impl Immutable for X` in the doc of a type and assume "its some immutable FP stuff or something"

Reviewer: Understood. IMO something that at least mentions immutability (e.g. `SharedImmutable`) would be preferable. I agree with @jswrenn that `Freeze` is an unintuitive name that most average programmers won't be familiar with, and it's confusing to have it conflated with the `freeze` operation.

Author: > `Freeze` is definitely not part of the average Rust programmer's lexicon — it's compiler jargon.

Nor will it really need to be, it's a rather niche feature; but people in that niche have become familiar with it for a long time.

> `Freeze`, as noted in the preceding paragraph, conflicts with an entirely unrelated concept.

An unrelated concept which it doesn't describe any better, and which is just as niche (in the sense that one has to be quite familiar with LLVM's variable model to understand it at all), and could possibly be expressed better in the standard library through APIs on `MaybeUninit` that would rely on this intrinsic.

Also, there's no reason to believe this conflict is all that important: they are 2 concepts in different namespaces that happen to share a name. One is a method and the other a trait; both of which requiring enough homework to use properly that I wouldn't expect anyone with a need for them to mistake one for the other.

> Naming is typically finalized during stabilization, and I'm surprised that this feature — unlike all other library additions (afaik) — is considered so urgent as to warrant a fast-tracked stabilization that bypasses the usual naming discussions.

Note that this is a stabilization PR, and that it's been up for long enough that I wouldn't quite call that "fast-tracked".

> Naming this trait is the one shot we get at picking what term should be in the average Rust programmer's lexicon, and I don't think we should discard that opportunity lightly.

I don't think it was discarded "lightly": it has been the object of much discussion, including the 2 design meetings on the topic of stabilization. It was discarded because no better name was found for it.

Reviewer 4: > Some languages are built around the concept of immutability, so it's not far fetch that someone may find `impl Immutable for X` in the doc of a type and assume "its some immutable FP stuff or something"

Note that `Freeze` carries this same pitfall. [In Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze), "freeze" is an operation that makes an object immutable, even if you're the sole owner of it.

Reviewer 13: For one of the earlier discussions, see:

- https://github.com/rust-lang/rust/issues/131401

For the rationale from @p-avital's design meeting document today, see:

> The most bikesheddy aspect of this RFC has been the question of renaming the trait, as it clashes with LLVM's `freeze` intrinsic.
>
> I don't believe the clash to be all that relevant: LLVM's `freeze` (which is itself jargoney) could become a method of `MaybeUninit`, with namespacing providing sufficient distinction between the features. Renaming the trait could also cause confusion when approaching documents that mention `Freeze`, as I doubt they'd all get updated; and `Freeze` has been a part of the Rust jargon for a few years now.
>
> While debating this, it's been generally agreed that the two properties that a new name _should_ hint at are:
> * The implication of immutability, as this is the core property;
> * The restriction of that immutability to non-indirected memory, as not hinting at it could lead to great (and possibly costly to someone) confusion. I have therefore omitted `Immutable` from the list of options.
>
> Here is a list of propositions collected as exhaustively as possible. My opinions on each will be surrounded by braces to keep a more compact style:
> - `ShallowImmutable` (a popular option, it fulfills both roles; though "shallow" may be confused for a negation of "nested", this ambiguity should, if anything, lead to a higher likelihood of people looking to disambiguate by reading the docs)
> - `LocalImmutable` (cut from the same cloth as `ShallowImmutable`)
> - `InlineImmutable` (again, similar structure)
> - `CellFree` (may be a bit too focused on implementation)
> - `ValueImmutable` (I dislike that one as it may get mistaken for "this is immutable unless indirected", or even "this value cannot be mutated, ever")
> - `DirectImmutable` (like `ShallowImmutable`, once more)
> - `InteriorImmutable` (while less eggregious than `Immutable`, it's the most verbose option yet _while_ not hinting at the restriction at all).
>
> An additional argument in favour of `Freeze` is the general convention that traits are named after verbs (eg. `Clone` and not `Cloneable`): conceptually, `Freeze` allows the obtention of a "frozen reference" which guarantees that the memory it points to cannot be modified as long as the reference lives, not even through interior mutability:
>
> ```rust
> trait Freeze {
>     /// Returns a `&frozen` reference to `self`, which is distinct
>     /// from a common reference in that the bytes from `&self as *const Self`
>     /// to `(&self as *const Self).add(1)` are guaranteed to stay the
>     /// same value for the entire lifetime of the frozen reference.
>     fn freeze(&self) -> &frozen Self;
> }
> ```
>
> This conceptual method is simply omitted because `T: Freeze` already provides the above stated guarantee for `&T`, and therefore `&T` and `&frozen T` are identical for `T: Freeze`.
>
> While I still prefer keeping `Freeze` to all of those options, on the grounds that it has already settled in the Rust jargon and that the suggested names haven't seemed to spark joy either during debates, I'd emphasize that apart from those I have justified a dislike for, I'd happily support any name that gets `Freeze` to finally get stabilized.

> - and that stabilizing this feature too valuable to hold it back on naming.

Reviewer 7: ```suggestion
- and that stabilizing this feature is too valuable to hold it back on naming.
```

- Feature Name: `stabilize_marker_freeze`
- Start Date: 2024-05-10
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Stabilize `core::marker::Freeze` in trait bounds.

# Motivation
[motivation]: #motivation

> In some contexts, guaranteeing that a value cannot be modified is a requirement to a construct behaving properly (`const` references, keys of a structured map...).

Reviewer 3: It's not clear to me how this helps for "keys of a structured map". Consider that `Box<T>: Freeze` for all `T` -- there can be interior mutability behind pointer indirection even with `Freeze`.

`Freeze` is entirely a by-value, "shallow" property.

Author: damn, that makes it way less useful for other purposes than static promotion then :(

Author: As an aside, could changing this shallowness property be an interesting direction for this RFC, or is the current shallowness important in other aspects at the moment?

Reviewer 3: It's crucial for it to remain shallow. That's the entire point the trait was introduced for: determine whether shared references permit mutation or not. In an `&Box<Cell<T>>`, the data behind the first pointer is fully immutable and hence the `Box` should be `Freeze`.

While this looks achievable by only exposing immutable references, this is actually insufficient due to the (necessary) existence of interior mutability.

Internally, the compiler uses `core::marker::Freeze` to identify types that are known not to have interior mutability.

This RFC seeks to stabilize this trait for trait bounds for the following reasons:
- As explained above, some existing constructs are left to "trust" that the values they'd need to freeze won't be mutated through interior mutability. This is a potential bug source for these constructs.
- With [this PR](https://github.com/rust-lang/rust/issues/121250), a breaking change was introduced (and stabilized in 1.78):
	- This change prevents associated consts from containing references to values that don't implement `core::marker::Freeze`. Since this bound cannot be added to generics, this prevents associated consts from containing references to generics altogether.

> 	- This change was done in order to prevent potential unsoundness, but it also completely prevents sound uses, such as the ones reported in [this regression issue](https://github.com/rust-lang/rust/issues/123281).

Reviewer 3: If there's a small self-contained example demonstrating the problem, it would be good to include that here.

RFCs should be self-contained documents motivating the desired change.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

This RFC simply seeks to act as the stabilization RFC for `core::marker::Freeze` in trait bounds.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The work necessary for this RFC has already been done and merged in [this PR](https://github.com/rust-lang/rust/issues/121675), and a [tracking issue](https://github.com/rust-lang/rust/issues/121675) was opened.

> However, it was deemed that adding an Auto Trait to Rust's public API should go through the proper RFC process, but the RFC was never created.

Reviewer 3: The RFC should spell out the drawbacks of a new auto trait more explicitly.

Reviewer 3: Specifically, auto traits are a semver hazard: adding a new private field to a type that already has private fields is a breaking change if that field removes the `Freeze` property of the surrounding type.

However, at least under some conditions this is already a breaking change today: if there are any constants `C` of a type `T`, then `const { &C }` will compile if `T: Freeze`, but may fail to compile otherwise. By exposing `Freeze` bounds it is now always a breaking change, even if there are no constants.

The [semver guide](https://doc.rust-lang.org/cargo/reference/semver.html) doesn't seem to discuss auto traits at all, or did I miss the relevant section @ehuss?

Reviewer 14: It hasn't been added yet. Many missing items are tracked in https://github.com/rust-lang/cargo/issues/8736, including "Adding or removing private fields of a struct can be a breaking change if those fields change the auto traits.". I'm sure there are many more.

# Drawbacks
[drawbacks]: #drawbacks

- Some people have previously argued that this would be akin to exposing compiler internals.
	- The RFC author disagrees, viewing `Freeze` in a similar light as `Send` and `Sync`: a trait that allows soundness requirements to be proven at compile time.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> - This trait has existed for 7 years. Over this course, it was made a compiler internal temporarily, but its usefulness lead to it being re-exposed soon after.

Reviewer 3: It was internal for several years. I wouldn't call that "soon after".

> - This trait has existed for 7 years. Over this course, it was made a compiler internal temporarily, but its usefulness lead to it being re-exposed soon after.

Reviewer 15: ```suggestion
- This trait has existed for 7 years. Over this course, it was made a compiler internal temporarily, but its usefulness led to it being re-exposed soon after.
```

Reviewer 3: This sentence is factually wrong anyway. https://github.com/rust-lang/rust/issues/60715 was opened in 2019 and it took 4 years until Freeze got exposed again.

Author: It is, I messed up with my git blames :/

provided that it is behind an indirection (such as `Box<UnsafeCell<U>>`).
Notable `!Freeze` types are [`UnsafeCell`](core::cell::UnsafeCell) and its safe wrappers
such as the types in the [`cell` module](core::cell), [`Mutex`](std::sync::Mutex), and [atomics](core::sync::atomic).

> Any type which contains a `!Freeze` type without indirection is also `!Freeze`.

Reviewer 13: Is this true?  I would expect a type that contains a `!Freeze` type to be `?Freeze`, and that seems to be what `rustc` expects also:

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&code=%23%21%5Bfeature%28freeze%2C+freeze_impls%29%5D%0A%23%21%5Bfeature%28negative_impls%29%5D%0A%23%21%5Ballow%28internal_features%29%5D%0A%0Ause+core%3A%3A%7Bcell%3A%3AUnsafeCell%2C+marker%3A%3AFreeze%7D%3B%0A%0Astruct+S%28UnsafeCell%3C%28%29%3E%29%3B%0A%2F%2Fimpl+%21Freeze+for+S+%7B%7D+%2F%2F+Uncomment+to+make+%60S%3A+%21Freeze%60.%0A%0Atrait+Tr%3CT%3E+%7B%7D%0A%0Aimpl+Tr%3CS%3E+for+%28%29+%7B%7D%0Aimpl%3CT%3A+Freeze%3E+Tr%3CT%3E+for+%28%29+%7B%7D%0A%2F%2F~%5E+ERROR+conflicting+implementations+of+trait+%60Tr%3CS%3E%60+for+type+%60%28%29%60%0A%0Afn+main%28%29+%7B%7D%0A)

If this should be changed, it'd be worth checking elsewhere in the document for other cases like this.

Reviewer 12: ```suggestion
Any type which contains a non-`Freeze` type without indirection also does not implement `Freeze`.
```

Yeah, if negative impls are part of the reasoning, then "a type that does not implement `Freeze`"/"a non-`Freeze` type" does not mean the same thing as "a `!Freeze` type", and I think the intent of this section is that the container type is the former. The contained type could be either equivalently though, I think[^1].

Also, `Mutex` is described above as a "notable `!Freeze` type", though there is currently no explicit negative impl for it (`Mutex` is "a type not implementing `Freeze`", but not "a type implementing `!Freeze`" according to negative-impls logic). Perhaps that section could instead say "Notable non-`Freeze` types are [...]"

[^1]: since `Freeze` is an auto trait, so any "non-`Freeze` type" either has an explicit `impl !Freeze`, or contains such a type.

-----

Note that using `Send` or `Sync` in place of `Freeze` in the playground snippet gives the same error, but `Clone` does not, which shows that this is probably `auto`-trait-specific logic somehow.

Author: @zachs18 thanks for clearing up the section. Indeed me using `!Freeze` as a shortcut for non-`Freeze` may have been misleading :)

One could argue that `UnsafeCell` and atomics _should_ be `!Freeze`, but that `Mutex` could be simply non-`Freeze` by virtue of the data it contains being inlined being an implementation detail

in code that holds an immutable reference to `T`.

# Semver hazard
`Freeze` being an auto-trait that encodes a low level property of the types it is implemented for, 
it is strongly advised against to rely on external types maintaining that property unless that
contract is explicitly stated out-of-band (through documentation, for example).

Conversely, authors that consider `Freeze` to be part of a type's contract should document this
fact explicitly.

## The ZST caveat
While `UnsafeCell<T>` is currently `!Freeze` regardless of `T`, allowing `UnsafeCell<T>: Freeze` if `T` is
a Zero-Sized-Type is currently under consideration.

Therefore, the advised way to make your types `!Freeze` regardless of their actual contents is to add a 
[`PhantomNotFreeze`](core::marker::PhantomNotFreeze) field to it.

[`PhantomData<T>`](core::marker::PhantomData) only implements `Freeze` if `T` does, making it a good way

> to conditionally remove a generic type's `Freeze` auto-impl.

Reviewer 3: This is currently not true (`PhantomData<T>` is always `Freeze`). And when I tried changing that there was significant fallout, though I do not remember the details.

Author: Arguably, this is an example of what the `Freeze` doc could/should look like _once implemented_ :p

# Safety
This trait is a core part of the language, it is just expressed as a trait in libcore for
convenience. Do *not* implement it for other types.
```

Mention could be added to `UnsafeCell` and atomics that adding one to a previously `Freeze` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoke its implementation of `Freeze`.

## Fixing `core::marker::PhantomData`'s `Freeze` impl

At time of writing, `core::marker::PhantomData<T>` implements `Freeze` regardless of whether or not `T` does.

This is now considered a bug, with the corrected behaviour being that `core::marker::PhantomData<T>` only implements `Freeze` if `T` does.

While crates that would "observe" this change exist, the current consensus is that it would only break invalid usages of that invariable bound.
Most crates that would observe this change could replace their usage of `core::marker::PhantomData<T>` by `core::marker::PhantomData<Ptr<T>>` where `Ptr<T>`
is a pointer-type with the relevant caracteristics regarding lifetime, `Send` and `Sync`ness.

> The author doesn't have the necessary knowledge to implement this change, which should still be subject to a crater run to ensure no valid use-cases were missed.

Reviewer 3: The change is trivial, see https://github.com/rust-lang/rust/pull/114559. The hard part is dealing with all the ecosystem breakage.

- Adding a `trait Pure: Freeze` which extends the interior immutability guarantee to indirected data could be valuable:
	- This is however likely to be a fool's errand, as indirections could (for example) be hidden behind keys to global collections. 
	- Providing such a trait could be left to the ecosystem unless we'd want it to be an auto-trait also (unlikely).

> - `impl !Freeze for T` could be a nice alternative to `PhantomNotFreeze`. However, this would be a significant language change that needs much deeper consideration.

Reviewer 13: It's worth mentioning that these are not the same, and so one may not be a straightforward alternative for the other.

Including a `PhantomNotFreeze` field removes the `Freeze` impl for the type.  `impl !Freeze for T` expresses a guarantee that there will *never* be a `Freeze` bound for the type.

Author: Right you are, if we stick to the "`impl !Trait for X` means that `impl Trait for X` requires a major bump" interpretation (which is something I'd really enjoy to have), they do indeed yield different results
